### ISSUE TITLE
Recursive rendering needs work

### ISSUE THREAD

**Issue by julienkosinski** [swagger-api/swagger-ui#3325]
Hello,

- swagger-ui version 3.0.17
- a swagger file reproducing the issue : https://gist.githubusercontent.com/julienkosinski/287a3b77e25a6ca3e6f7a0293bea4e3d/raw/fd94d4f1457ccbaa538f27996dba5d077d9c6f36/gistfile1.txt

I have the following model definitions spec:

```
    "definitions": {
        "models.Equipment": {
            "title": "Equipment",
            "type": "object",
            "properties": {
                "Features": {
                    "type": "array",
                    "items": {
                        "$ref": "#/definitions/models.Feature"
                    }
                },
                "Id": {
                    "type": "integer",
                    "format": "int64"
                },
                "IdType": {
                    "type": "string"
                },
                "Name": {
                    "type": "string"
                },
                "Price": {
                    "type": "integer",
                    "format": "int32"
                }
            }
        },
        "models.Feature": {
            "title": "Feature",
            "type": "object",
            "properties": {
                "Equipments": {
                    "type": "array",
                    "items": {
                        "$ref": "#/definitions/models.Equipment"
                    }
                },
                "Id": {
                    "type": "integer",
                    "format": "int64"
                },
                "IdFeature": {
                    "$ref": "#/definitions/models.Feature"
                },
                "Name": {
                    "type": "string"
                }
            }
        }
    }
```

In the `Feature` model, the `Equipments` property is defined as an array of `Equipment` models, but Swagger UI 3.x renders it as an empty array `[]`. Everywhere `Feature` model is used, like as examples for `POST` method in `Feature` I have this kind of display.

![swagger-ui bug](https://i.stack.imgur.com/MBgUS.png)

We think it may be a bug caused by circular references. Thanks to [Helen](https://stackoverflow.com/questions/44871128/object-array-rendered-as-empty-array-in-swagger-ui).

Thank you very much!

**Comment by julienkosinski** [comment]
@shockey @webron I mention you just to make sure this has been noticed, don't be offended :).

**Comment by shockey** [comment]
@julienkosinski thanks, I'll get this triaged on Monday 😄

**Comment by julienkosinski** [comment]
@shockey Thank you! Yeah, of course, have a nice weekend! :)

**Comment by shockey** [comment]
I'm seeing two things here:

1. `Feature -> Equipments` is being passed a $ref to render... this shouldn't be happening.
2. The nested model is showing a very strange generated name:

![image](https://user-images.githubusercontent.com/680248/28295530-11d9e664-6b16-11e7-9450-428f12e54014.png)


This isn't quite right - we'll look into it 😄

**Comment by julienkosinski** [comment]
Thank you very much for your time! :D

**Comment by kverfaille** [comment]
In case it helps to find the issue => This indeed seems to be "Circular references" ... it also seems to appear without array wrapping.

```
"definitions": {
   "Classification": {
            "title": "Classification",
            "properties": {
                "ID": {
                    "type": "integer",
                    "default": 126132
                },
                "rank": {
                    "type": "string"
                },
                "scientificname": {
                    "type": "string"
                },
                "child": {
                    "$ref": "#/definitions/Classification"
                }
            },
            "xml": {
                "name": "Classification",
                "wrapped": false
            }
        }
}
```
Results in:
![image](https://user-images.githubusercontent.com/11004408/28569303-890977bc-7139-11e7-8624-dc53428c6822.png)

In the 2.x versions this was handled as follows:
![image](https://user-images.githubusercontent.com/11004408/28569465-2e840018-713a-11e7-88ae-d230c17eff0b.png)

Further the Model example values (JSON and XML) don't handle these nested definitions. The property is just removed from the list. But this could be another issue.

![image](https://user-images.githubusercontent.com/11004408/28569650-08327664-713b-11e7-8a4b-b4d6684cfbba.png)

**Comment by tobymurray** [comment]
Just ran into this as well using the OpenAPI 3 spec:
```
{
  "openapi": "3.0.0",
  "info": {
    "title": "Circular reference example",
    "version": "0.1"
  },
  "paths": {},
  "components": {
    "schemas": {
      "SelfReferencingSchema": {
        "type": "object",
        "properties": {
          "children": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/SelfReferencingSchema"
            }
          }
        }
      }
    }
  }
}
```

Yields the following:
![image](https://user-images.githubusercontent.com/3683198/31150420-7bcaa4cc-a861-11e7-8d9f-9ae058d3ed55.png)

I'm not familiar with the history or previous implementations of this or what's possible/hard, but personally I think it'd be nice to show the title for a circular reference but not expand it by default. E.g.:
![image](https://user-images.githubusercontent.com/3683198/31150518-e31cbe4e-a861-11e7-9180-0fad8c8ccece.png)

**Comment by Rinzwind** [comment]
Here's a screenshot of another example I was testing. The structure of the schema object for the “containingFolder” attribute for “Folder” is similar to the one for “resourceFork” for “File”; I’m guessing the reason that it’s not shown that “containingFolder” is another “Folder” is due to the circularity.

![swagger_ui](https://user-images.githubusercontent.com/1611248/31171070-ca492108-a8fe-11e7-9bb6-c050bc7ad694.png)

Here’s the OpenAPI object:

```
{
  "openapi": "3.0.0",
  "info": {
    "title": "Test",
    "version": "0.1"
  },
  "components": {
    "schemas": {
      "Fork": {
        "type": "object",
        "properties": {
          "data": {
            "type": "string",
            "format": "base64"
          }
        }
      },
      "File": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "dataFork": {
            "$ref": "#/components/schemas/Fork"
          },
          "resourceFork": {
            "anyOf": [
              {
                "type": "null"
              },
              {
                "$ref": "#/components/schemas/Fork"
              }
            ]
          },
          "containingFolder": {
            "$ref": "#/components/schemas/Folder"
          }
        }
      },
      "Folder": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "containingFolder": {
            "anyOf": [
              {
                "type": "null"
              },
              {
                "$ref": "#/components/schemas/Folder"
              }
            ]
          },
          "containedItems": {
            "type": "array",
            "items": {
              "anyOf": [
                {
                  "$ref": "#/components/schemas/File"
                },
                {
                  "$ref": "#/components/schemas/Folder"
                }
              ]
            }
          }
        }
      }
    }
  }
}
```

**Comment by hkosova** [comment]
@Rinzwind

```
          "resourceFork": {
            "anyOf": [
              {
                "type": "null"
              },
              {
                "$ref": "#/components/schemas/Fork"
              }
            ]
```
`"type": "null"` is not valid in OpenAPI, because [there's no `null` type](https://swagger.io/docs/specification/data-models/data-types/#null) - that's one of the differences from JSON Schema. OpenAPI uses the `nullable` attribute instead.

I'm not sure if there's a way to combine `$ref` with `nullable`. You'll probably need to add `nullable: true` directly to the referenced schemas (Fork and Folder).

**Comment by Rinzwind** [comment]
> "type": "null" is not valid in OpenAPI, because there's no null type - that's one of the differences from JSON Schema. OpenAPI uses the nullable attribute instead.

@hkosova, I had missed that, thanks for pointing it out!

> I'm not sure if there's a way to combine `$ref` with `nullable`. You'll probably need to add `nullable: true` directly to the referenced schemas (Fork and Folder).

I’m not sure that adding `nullable` directly to the referenced schemas is going to match with what I was trying to express: for the references to Fork from File, only the value for “resourceFork” can be `null`, the value for “dataFork” can not be `null`. I guess I could still express this as follows, which as far as I can tell is at least in accordance with the OpenAPI specification? But using an “anyOf” with just one containing schema seems clumsy, is there a better way to handle this?

```
... 
"File": {
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "dataFork": {
      "$ref": "#/components/schemas/Fork"
    },
    "resourceFork": {
      "nullable": true,
      "anyOf": [
        {
          "$ref": "#/components/schemas/Fork"
        }
      ]
    },
    ...
```

**Comment by webron** [comment]
@Rinzwind can't think of a different way to express it.

**Comment by Rinzwind** [comment]
> @Rinzwind can't think of a different way to express it.

Thanks for the feedback. I’ve posted an [issue for it at the OpenAPI Specification repository](https://github.com/OAI/OpenAPI-Specification/issues/1368).

**Comment by jonschoning** [comment]
Could you also use `allOf` , supported by OAS 2.0?  If so, would there be a difference between using the `anyOf`` method?
```
... 
"File": {
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "dataFork": {
      "$ref": "#/components/schemas/Fork"
    },
    "resourceFork": {
      "allOf": [
        {
          "$ref": "#/components/schemas/Fork"
        },
        {
          "nullable": true
        }
      ]
    },
    ...
```

**Comment by webron** [comment]
`allOf` would actually be better than `anyOf` in this case.

**Comment by Rinzwind** [comment]
@webron: I am not sure whether the [example given by @jonschoning](https://github.com/swagger-api/swagger-ui/issues/3325#issuecomment-339396192) has the intended semantics.

The [section on “allOf” in the JSON Schema Validation proposal](https://tools.ietf.org/html/draft-wright-json-schema-validation-00#section-5.22) says: “An instance validates successfully against this keyword if it validates successfully against all schemas defined by this keyword's value.”

As far as I understand, `null` validates successfully against the schema `{ "nullable": true }`. But `null` does *not* validate successfully against the schema `{ "$ref": "#/components/schemas/Fork" `}. Therefore, I take it `null` also does *not* validate successfully against this schema:

```
{
  "allOf": [
    {
      "$ref": "#/components/schemas/Fork"
    },
    {
      "nullable": true
    } ] }
```

In other words, in @jonschoning's example, `null` would *not* be a valid value for “resourceFork”. Or did I misunderstand how schemas work in OpenAPI?

**Comment by jonschoning** [comment]
@rinzwind at the same time that allOf is interpret the ref is dereferenced, so the net effect results in the nullable being a part of the refs definition.  You can check this understanding by considering how allOf works with appending properties to a ref with allOf

**Comment by webron** [comment]
Actually, @Rinzwind's point is correct, and that won't work. However, using `anyOf` also means that any type can be used because of nullable only schema.

**Comment by jonschoning** [comment]
@webron I have a hard time finding a working example of how nullable works https://github.com/OAI/OpenAPI-Specification/search?utf8=%E2%9C%93&q=nullable&type=

**Comment by webron** [comment]
If you find that an example is missing, feel free to open an issue on that repo. Not really sure what's not clear about it though.

**Comment by jonschoning** [comment]
I find @Rinzwind 's interpretation curious.  if you look the following valid schema (that does not use nullable)
```
paths:
  /feature/:
    get:
      description: get
      operationId: get
      responses:
        '200':
          description: none
          schema:
            allOf:
              - $ref: '#/definitions/models.Feature'
              - properties:
                  Age:
                    type: integer
definitions:
  models.Feature:
    title: Feature
    type: object
    properties:
      Name:
        type: string
```
then following @Rinzwind 's logic :

As far as I understand, `{ Name: "foo"}` validates successfully against the schema `$ref: '#/definitions/models.Feature'`. But `{ Name: "foo"}` does not validate successfully against the schema `properties: Age: type: integer`. Therefore, I take it `{ Name: "foo"}` also does not validate successfully against the above schema.

Which is incorrect logic because the above schema is valid.  Therefore something is not clear about that logic as it pertains to how `allOf` combines with the other features

**Comment by Rinzwind** [comment]
@jonschoning As far as I understand, the following is meant by “Allows sending a `null` value for the defined schema. Default value is false.”: “When `true`, allows sending any value; when not present or `false`, allows sending any value except `null`.”

Rewording the definition according to the terminology employed in JSON Schema Validation (using in particular the wording for [“uniqueItems”](https://tools.ietf.org/html/draft-wright-json-schema-validation-00#section-5.12) as inspiration):

> **nullable**
>
> The value of this keyword MUST be a boolean.
>
> If this keyword has boolean value `true`, any instance validates successfully. If this keyword has  boolean value `false`, any instance validates successfully unless it is `null`.
>
> If not present, this keyword MUST be considered present with boolean value `false`.

I hope that helps. Keep in mind though that this is not a direct quote from the specification, but just my understanding of it.

**Comment by jonschoning** [comment]
Ok, I understand how, thank you.

It seems one cannot simply specify a ref should be nullable then with anyOf or allOf without also changing what validates per Ron's comment

**Comment by Rinzwind** [comment]
> It seems one cannot simply specify a ref should be nullable then with anyOf or allOf without also changing what validates per Ron's comment

Indeed, as far as I understand, a schema like `{"anyOf":[...,{"nullable":"true"}]}` would allow `null`, but also *any* other value. The schema `{"nullable":"true"}` in OpenAPI is not equivalent to the schema `{"type":"null"}` in (pure) JSON Schema. I find the latter easier to understand, I'm not sure why OpenAPI introduced the former. I think I intuitively grasp its meaning, but [as mentioned above](https://github.com/swagger-api/swagger-ui/issues/3325#issuecomment-339811553), I'm having some trouble giving an exact definition for it. I agree with [your earlier comment](https://github.com/swagger-api/swagger-ui/issues/3325#issuecomment-339716376) that the OpenAPI specification isn't very clear about the definition of “nullable”.

**Comment by webron** [comment]
It looks like the original issue has been resolved, but there's still work to do on recursive rendering. The ticket has been renamed to reflect that.

**Comment by Handled1** [comment]
> It looks like the original issue has been resolved, but there's still work to do on recursive rendering. The ticket has been renamed to reflect that.

Hi, can I just check in to see if anyone has been working on this?  I see the recursive rendering still not working in swagger

**Comment by lorthirk** [comment]
+1. I have the following object:

```
account:
  type: object
  allOf:
    - $ref: '../openapi.yaml#/components/schemas/baseEntity'
    - properties:    
        childAccounts:
          type: array
            items:
              $ref: '#/components/schemas/
```

This causes an error in Swagger-UI:

![image](https://user-images.githubusercontent.com/8140139/64597224-a0185000-d3b5-11e9-9b8a-b9073204c559.png)

```
index.js:2247 TypeError: Cannot read property '1' of undefined
    at l (core.js:244)
    at Object.p [as applyPatch] (core.js:277)
    at Object.applyPatch (index.js:940)
    at e.value (index.js:2258)
    at index.js:2242
    at Array.forEach (<anonymous>)
    at e.value (index.js:2214)
    at index.js:2279
    at async Promise.all (:5001/index 0)
(anonymous) @ index.js:2247
value @ index.js:2214
(anonymous) @ index.js:2279
Promise.then (async)
value @ index.js:2276
(anonymous) @ index.js:2232
value @ index.js:2214
a @ index.js:2412
(anonymous) @ index.js:2390
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
(anonymous) @ index.js:2406
value @ index.js:2380
v @ index.js:2419
Rt @ index.js:2707
(anonymous) @ index.js:2802
c @ runtime.js:45
(anonymous) @ runtime.js:271
e.<computed> @ runtime.js:97
o @ asyncToGenerator.js:5
s @ asyncToGenerator.js:27
(anonymous) @ asyncToGenerator.js:34
t @ _export.js:36
(anonymous) @ asyncToGenerator.js:23
Ft @ index.js:2813
In.resolveSubtree @ index.js:2777
resolveSubtree @ index.js:22
(anonymous) @ actions.js:178
c @ runtime.js:45
(anonymous) @ runtime.js:271
e.<computed> @ runtime.js:97
o @ asyncToGenerator.js:5
s @ asyncToGenerator.js:27
Promise.then (async)
o @ asyncToGenerator.js:15
s @ asyncToGenerator.js:27
(anonymous) @ asyncToGenerator.js:34
t @ _export.js:36
(anonymous) @ asyncToGenerator.js:23
(anonymous) @ actions.js:176
(anonymous) @ actions.js:176
c @ runtime.js:45
(anonymous) @ runtime.js:271
e.<computed> @ runtime.js:97
o @ asyncToGenerator.js:5
s @ asyncToGenerator.js:27
(anonymous) @ asyncToGenerator.js:34
t @ _export.js:36
(anonymous) @ asyncToGenerator.js:23
b @ debounce.js:95
x @ debounce.js:144
w @ debounce.js:132
setTimeout (async)
w @ debounce.js:135
setTimeout (async)
(anonymous) @ debounce.js:103
E @ debounce.js:172
(anonymous) @ actions.js:243
(anonymous) @ utils.js:134
(anonymous) @ bindActionCreators.js:3
(anonymous) @ OperationContainer.jsx:155
value @ OperationContainer.jsx:89
e.notifyAll @ CallbackQueue.js:74
close @ ReactReconcileTransaction.js:78
closeAll @ Transaction.js:207
perform @ Transaction.js:154
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
w @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
e @ ReactUpdates.js:198
a @ ReactUpdateQueue.js:22
enqueueSetState @ ReactUpdateQueue.js:216
s.setState @ ReactBaseClasses.js:62
i.handleChange @ connect.js:302
f @ createStore.js:172
(anonymous) @ utils.js:137
(anonymous) @ bindActionCreators.js:3
(anonymous) @ wrap-actions.js:9
r @ system.js:174
(anonymous) @ system.js:461
(anonymous) @ index.js:22
r @ system.js:174
(anonymous) @ system.js:461
(anonymous) @ actions.js:77
(anonymous) @ utils.js:134
(anonymous) @ bindActionCreators.js:3
(anonymous) @ wrap-actions.js:5
r @ system.js:174
(anonymous) @ system.js:461
(anonymous) @ index.js:11
r @ system.js:174
(anonymous) @ system.js:461
p @ download-url.js:37
Promise.then (async)
(anonymous) @ download-url.js:26
(anonymous) @ utils.js:134
(anonymous) @ bindActionCreators.js:3
h @ index.js:153
Vn @ index.js:180
window.onload @ (index):42
load (async)
(anonymous) @ (index):39
Show 52 more frames
system.js:464 TypeError: Cannot read property 'every' of undefined
    at actions.js:191
    at reducers.js:93
    at immutable.js:3084
    at ft.__iterate (immutable.js:2206)
    at o.__iterateUncached (immutable.js:3083)
    at ce (immutable.js:604)
    at o.K.__iterate (immutable.js:320)
    at o.forEach (immutable.js:4381)
    at immutable.js:2069
    at ft.Ue.withMutations (immutable.js:1353)
```

As soon as I comment out that property, no errors in the UI.

**Comment by Darkzarich** [comment]
+1. I use this definition:

```
"Comment": {
        "type": "object",
        "properties": {
          "id": {
            "type": "string"
          },
          "post": {
            "type": "string"
          },
          "body": {
            "type": "string"
          },
          "parent": {
            "type": "string"
          },
          "author": {
            "$ref": "#/components/schemas/Author"
          },
          "createdAt": {
            "type": "string",
            "example": "2019-08-16T01:04:02.504Z"
          },
          "children": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/Comment"
            }
          },
          "rating": {
            "type": "number"
          },
          "rated": {
            "$ref": "#/components/schemas/UserRate"
          }
        }
      },
```
`children` is null in response body.

**Comment by LaysDragon** [comment]
Encounter the same problem of recursive data structure...swagger-ui just don't rendering it properly.
Is there any plan to fix this?

**Comment by Phobossus** [comment]
Bump.

**Comment by mabzzz** [comment]
+1

**Comment by zorankabic** [comment]
+1

**Comment by ChinmayanP** [comment]
Is there any update on this?

**Comment by manmoth** [comment]
+1 Seems to be related, but with different errors in UI (Swashbuckle.AspNetCore.SwaggerUI 5.0.0):
![image](https://user-images.githubusercontent.com/944769/75546190-50434900-5a28-11ea-9605-fea72ed9c179.png)

{
  "swagger": "2.0",
  "info": {
    "description": "Version 1",
    "version": "v1"
  },
  "paths": {
    "/Recursive": {
      "get": {
        "tags": [
          "Recursive"
        ],
        "produces": [
          "text/plain",
          "application/json",
          "text/json"
        ],
        "responses": {
          "200": {
            "description": "Success",
            "schema": {
              "$ref": "#/definitions/RecursiveType"
            }
          }
        }
      }
    }
  },
  "definitions": {
    "RecursiveType": {
      "type": "object",
      "properties": {
        "item": {
          "allOf": [
            {
              "$ref": "#/definitions/RecursiveType"
            }
          ],
          "readOnly": true
        }
      }
    }
  }
}

**Comment by Cybrosys** [comment]
Hello,

I just encountered this issue as well. I have a list of Item instances. Each Item instance in that list has a property that's also a list of Item instances; a recursive data structure. Swagger does not like this.

Is there something that can be done for now, besides excluding the property that is?

**Comment by aaaee2** [comment]
Not the most elegant solution, but it is possible to show the user what type this property is.

```yaml
  schemas:
    UserView:
      type: object
      properties:
        id:
        # ....
        # ....
        # ....
        master:
          $ref: '#/components/schemas/UserViewCircular'
      required:
        # ....
    UserViewCircular:
      $ref: '#/components/schemas/UserView'
```
![Screenshot_1](https://user-images.githubusercontent.com/15371002/77225067-92b5fc80-6b74-11ea-8a01-6c7af3fee7c3.png)

![Screenshot_2](https://user-images.githubusercontent.com/15371002/77225115-f3453980-6b74-11ea-8d93-6847632d0686.png)

**Comment by inet-cologne** [comment]
There are also render errors in swagger hub swagger-editor.
I know that I've a recursive reference in my (quite complex) schema and it took me a long time to recognize that the UI error seems to come from there (see screenshot):
<img width="906" alt="swagger-ui-render-error-recursive-references_2020-03-31" src="https://user-images.githubusercontent.com/42293052/77969604-57d96480-72ea-11ea-8ae6-9681880bebd6.png">

If I click randomly on some schema components in the left navigation bar, on nearly every click the editor increases the error count just saying the same error again and again: "error, line 0, undefined is not an object (evaluating 'm[b]')".

Hope there will be soon a fix for that available.
Thanks in advance,
Chris

**Comment by serieznyi** [comment]
I thin we can do something like this

```yaml
components:
  schemas:
    CategoryWithReferences:
      allOf:
        - $ref: '#/components/schemas/Category'
        - type: object
          properties:
            parent:
              allOf:
                - $ref: '#/components/schemas/Category'
            childs:
              type: array
              items:
                allOf:
                  - $ref: '#/components/schemas/Category'

    Category:
      type: object
      required:
        - id
        - name
      properties:
        id:
          type: integer
        name: 
          type: string

```

![image](https://user-images.githubusercontent.com/2323142/78915025-fc4e6a00-7a93-11ea-8146-9ae332717100.png)

**Comment by cem** [comment]
There's a very weird behavior with the call to SwaggerUIBundle. If the _url_ param is a absolute path, e.g. "/api/swagger.json", the JSON is loaded once (observed in devtools network monitor) and **circular references appear empty**. If the _url_ param is a relative path, e.g. "./swagger.json", the JSON is loaded twice and circular references are filled in!

**Comment by StenCalDabran** [comment]
I created a repository which shows the same problem in a specification generated from a NestJS application: https://codesandbox.io/s/nestjs-swagger-description-bug-dfesh?file=/src/app.controller.ts
OpenApi 3.0, no circular references, just recursive references.

Interestingly enough, the rendering problem only occurs within Responses -> Schema; if you look at the Schemas in the Schema summary at the bottom, everything is rendered correctly.
<details>
<summary>The used generated OpenApi spec document looks like this: </summary>
<p>

```JSON
{
  "openapi": "3.0.0",
  "info": {
    "title": "",
    "description": "",
    "version": "1.0.0",
    "contact": {}
  },
  "tags": [],
  "servers": [],
  "components": {
    "schemas": {
      "InnerType": {
        "type": "object",
        "properties": {
          "innerProperty": {
            "type": "number"
          },
          "innerPropertyWithDescription": {
            "type": "number",
            "description": "Inner description works."
          }
        },
        "required": ["innerProperty", "innerPropertyWithDescription"]
      },
      "OuterTypeWithDescription": {
        "type": "object",
        "properties": {
          "secondLevel": {
            "description": "Second Level level is faulty, it shows only this description instead of the InnerType.",
            "allOf": [
              {
                "$ref": "#/components/schemas/InnerType"
              }
            ]
          }
        },
        "required": ["secondLevel"]
      },
      "OuterTypeWithoutDescription": {
        "type": "object",
        "properties": {
          "secondLevel": {
            "$ref": "#/components/schemas/InnerType"
          }
        },
        "required": ["secondLevel"]
      },
      "ResponseType": {
        "type": "object",
        "properties": {
          "innerType": {
            "description": "First level works.",
            "allOf": [
              {
                "$ref": "#/components/schemas/InnerType"
              }
            ]
          },
          "faultyOuterType": {
            "description": "Outer description works.",
            "allOf": [
              {
                "$ref": "#/components/schemas/OuterTypeWithDescription"
              }
            ]
          },
          "workingOuterType": {
            "description": "Outer description works.",
            "allOf": [
              {
                "$ref": "#/components/schemas/OuterTypeWithoutDescription"
              }
            ]
          }
        },
        "required": ["innerType", "faultyOuterType", "workingOuterType"]
      }
    }
  },
  "paths": {
    "/test": {
      "get": {
        "operationId": "AppController_test",
        "parameters": [],
        "responses": {
          "200": {
            "description": "Response description works.",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ResponseType"
                }
              }
            }
          }
        }
      }
    }
  }
}
```
</p>
</details>

**Comment by yanalapraneetha** [comment]
Hi, 
I am facing the same issue with circular reference in my spec
### api spec
`openapi: 3.0.1
info:
  title: API
  version: 0.1.0
paths:
  /v1/message/{context}:
    get:
      tags:
      - cpsapi
      description: get pulse messages given the context in isolationId/modelId
      operationId: getPulseMessages
      parameters:
      - name: context
        in: path
        description: context id of message
        required: true
        schema:
          type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: '#/components/schemas/Message'
        "401":
          description: Not authenticated
        "500":
          description: Internal server error
components:
  schemas:
    Message:
      type: object
      properties:
        ID:
          type: string
          description: unique id for the message
        message:
          type: string
          description: the actual pulse message
        replies:
          type: array
          description: Get replies
          items:
            $ref: '#/components/schemas/Message'`

Reply property in above spec is shown in swagger ui like -
 "replies": [
      "string"
    ]

[message.txt](https://github.com/swagger-api/swagger-ui/files/6516214/message.txt)

Any update on this issue?

**Comment by vanDarg** [comment]
> 
> 
> I created a repository which shows the same problem in a specification generated from a NestJS application: https://codesandbox.io/s/nestjs-swagger-description-bug-dfesh?file=/src/app.controller.ts
> 
> OpenApi 3.0, no circular references, just recursive references.
> 
> Interestingly enough, the rendering problem only occurs within Responses -> Schema; if you look at the Schemas in the Schema summary at the bottom, everything is rendered correctly.
> The used generated OpenApi spec document looks like this:

Interesting detail about your example: I just took this to the Swagger Editor ([https://editor.swagger.io/](https://editor.swagger.io/)) to see if it would resemble a similar issue I am having. I noticed the 'components' part was written before the 'paths' part. When I switched the order of those two (parts before components) it would render correctly in both places.

**Comment by StenCalDabran** [comment]
> 
> 
> @StenCalDabran Interesting detail about your example: I just took this to the Swagger Editor (https://editor.swagger.io/) to see if it would resemble a similar issue I am having. I noticed the 'components' part was written before the 'paths' part. When I switched the order of those two (parts before components) it would render correctly in both places.

@vanDarg Very interesting indeed, thanks for pointing that out. I opened an issue in the nestjs/swagger repository (nestjs/swagger#1369), maybe it is possible to automatically generate the document in the other order in the first place - which of course would only mitigate the actual problem.

**Comment by ponelat** [comment]
Yeah, appears to still be an issue... created small reproduction based on example in first comment: https://gist.github.com/ponelat/e3e7e3e55247ef6d2cd09926bbeb4241

**Comment by benjamin-mogensen** [comment]
This is still an issue. I have a quite complex, deep schema, which Swagger UI reports it cannot resolve as soon as clicking on the model either under the endpoint or in the models section. Until I click those the UI looks fine. The $ref I have is recursive (a package can have inner packages AND can have an out package).
![image](https://user-images.githubusercontent.com/73802221/148501085-91deca5d-a15a-4cf6-9866-9b7e6a53d739.png)

**Comment by feenr** [comment]
This issue may have been fixed? I was going to attempt a fix, but checking out the latest swagger-ui it wasn't reproducible.

**Comment by StenCalDabran** [comment]
> This issue may have been fixed? I was going to attempt a fix, but checking out the latest swagger-ui it wasn't reproducible.

At least the example I built (see [my previous reply)](#issuecomment-824875828) still does not work with the swagger editor, which seems to use the latest swagger-ui. If the original issue is indeed resolved, I could open a new issue, but they seem to have similar root problems.

**Comment by julienkosinski** [comment]
Hello,
I don't monitor this issue anymore. If needed, I can close it now.

**Comment by gwcoffey** [comment]
This is still a problem in 4.5.0. Here's a minimal OAS that reproduces it. 

```json
{
  "openapi" : "3.0.1",
  "info" : {
    "title" : "Recursive Failure",
    "version" : "1.0"
  },
  "paths" : {
    "/test" : {
      "get" : {
        "responses" : {
          "200" : {
            "description": "Recursive example",
            "content" : {
              "application/json" : {
                "schema" : {
                  "$ref" : "#/components/schemas/test-response"
                }
              }
            }
          }
        }
      }
    }
  },
  "components" : {
    "schemas" : {
      "test-response": {
        "type": "object",
        "properties": {
          "foo": {
            "type": "string"
          },
          "bar": {
            "$ref" : "#/components/schemas/test-response"
          }
        }
      }
    }
  }
}
```

Here's what I see in swagger ui when I expand the `/test` path:

<img width="800" alt="Screen Shot 2022-02-16 at 11 35 17 AM" src="https://user-images.githubusercontent.com/149113/154342120-3bc4fefb-6c94-4068-b1b0-9c086885afd8.png">

Here's how the example is rendered:

```json
{
  "foo": "string"
}
```

And here's how the schema is rendered when expanded:

<img width="256" alt="Screen Shot 2022-02-16 at 11 37 21 AM" src="https://user-images.githubusercontent.com/149113/154342446-8719a63e-3f46-4854-af9e-53fc034f0686.png">

**Comment by geraldus** [comment]
Recursive schemas are still rendered incorrectly as `{}`. Note that there is no error appears in UI in my case, because generated `swagger.json` seems to be valid and have proper ref within.

**Comment by mristin** [comment]
Hi,
I'd like to have a try at fixing this as it is critical feature for us with many users. Does anybody know whether somebody else is working on this, or what a good entry point for this problem would be?

**Comment by RockyMM** [comment]
Just to put it outhere, some of the schemas from this bug report do render correctly, namely the initial bug report seems to be solved. Some are still not rendered correctly. I am using https://editor-next.swagger.io/ to test.

**Comment by atereques** [comment]
Hello,
I'm facing the same problem with schema rendering (and also with code completition, when creating JSON based on schema in IDE). We have custom query language, where filtering constraints can be nested within each other, see example of appropriate schema below. 
```
openapi: 3.0.1
info:
  title: Web service for catalogue Brand
  version: 1.0.0-oas3
paths:
  /api/rest/catalogs/brands:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Query'
      responses:
        '200':
          description: Query result
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Entity'
components:
  schemas:
    Entity:
      properties:
        primaryKey:
          type: integer
    Query:
      type: object
      properties:
        limit:
          type: integer
        filterBy:
          $ref: '#/components/schemas/FilterContainer'
    FilterContainer:
      type: object
      properties:
        and:
          type: array
          items:
            $ref: '#/components/schemas/FilterContainer'
        not:
          $ref: '#/components/schemas/FilterContainer'
        attribute_status_lessThan:
          type: string
        attribute_status_equals:
          type: string
```

**Comment by Hronom** [comment]
Hello, I'm facing similar issues like described here https://github.com/swagger-api/swagger-ui/issues/3325#issuecomment-662030081 is there solution for it?

**Comment by dtc-lan1wang** [comment]
still no solutions here?

**Comment by bergwerf** [comment]
This is still unsolved.

**Comment by Hronom** [comment]
+1 Same story and it's strange situation, when we use swagger-ui it's throw this error... BUT... when we use same swagger definition with this https://springdoc.org/ no error...

Please fix this, since it's create unnecessary questions/confusion for developers that use it.

I hope maintainers of this project very deep in context and able to find reasons for it as same as fix

**Comment by benjamin-mogensen** [comment]
In my experience Redoc and RapiDoc does not have this issue and also works with OAS 3.1 which swagger ui does not. So we use those instead.

**Comment by mbarmettler** [comment]
+1 we need to render inheritance structure like this:

public class TimelineHolidayQueryModel : TimeLineQuery
public class TimeLineQuery: PagedQuery
public class PagedQuery

after second inheritance, codegen to get Typescript interfaces, stops rendering

**Comment by s3tezsky** [comment]
Hi, I have run into the same issue (thanks to different lib using swagger-ui).

After some investigation I found out this issue was resolved in [v5.11.8](https://github.com/swagger-api/swagger-ui/releases/tag/v5.11.8).

**Comment by Lyumih** [comment]
It's steal actualy.
Please fix Recursive tree type scheme!

**Comment by medoni** [comment]
Any update on this?
